test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NetCorePackageReferenceProjectTests.cs

```diff
@@ -447,6 +456,1909 @@ public async Task GetInstalledVersion_WithAssetsFile_ChangingPackageSpec_Returns
+
+                        // This code added because nuGetPackageManager.InstallPackageAsync doesn't do updating ProjectSystemCache
+                        var installed = new LibraryDependency
+                        {
+                            LibraryRange = new LibraryRange(
+                                name: packageB.Id,
+                                versionRange: new VersionRange(packageB.Version),
+                                typeConstraint: LibraryDependencyTarget.Package),
+                        };
+
+                        var packageSpec = packageSpecs[i];
+                        packageSpec.TargetFrameworks.First().Dependencies.Add(installed);
+                        var projectRestoreInfo = ProjectJsonTestHelpers.GetDGSpecFromPackageSpecs(packageSpec);
+                        var projectNames = GetTestProjectNames(projectFullPaths[i], $"project{i}");
+                        projectCache.AddProjectRestoreInfo(projectNames, projectRestoreInfo, new List<IAssetsLogMessage>());
+                    }
+
+                    var initialInstalledPackages = await netCorePackageReferenceProjects[numberOfProjects-1].GetInstalledPackagesAsync(CancellationToken.None);
+
+                    var targetProjects = new List<NetCorePackageReferenceProject>()
+                    {
+                        netCorePackageReferenceProjects[numberOfProjects-1] // Top parent project.
+                    };
+
+                    // Act
+                    var results = await nuGetPackageManager.PreviewProjectsInstallPackageAsync(
+                        targetProjects,
+                        packageB_UpgradeVersion,
+                        resolutionContext,
+                        testNuGetProjectContext,
+                        sourceRepositoryProvider.GetRepositories(),
+                        CancellationToken.None);
+                    var actions = results.Select(a => a.Action).ToArray();
+                    await nuGetPackageManager.ExecuteNuGetProjectActionsAsync(
+                        netCorePackageReferenceProjects,
+                        actions,
+                        testNuGetProjectContext,
+                        new SourceCacheContext(),
+                        CancellationToken.None);
+
+                    // Assert
+                    Assert.Equal(initialInstalledPackages.Count(), 2);
+                    var builtIntegratedActions = actions.OfType<BuildIntegratedProjectAction>().ToList();
+                    Assert.Equal(actions.Count(), targetProjects .Count());
+                    Assert.Equal(actions.Count(), builtIntegratedActions.Count());
+                    // Uprade succeed for this top parent project(no parent but with childs).
+                    // Keep existing Upgrade/downgrade of individual project logic and making sure that my change is not breaking it.
+                    Assert.Equal(actions.Count(), builtIntegratedActions.Count());
```

Comment:
Similar here. 

Please validate the exact version of the expected package.

Reply:
Added version validate asserts.